{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Edge Addition Algorithm - simple implementation example"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"3\">Run time is around ~5 minutes with the default input. \\\n",
    "\\\n",
    "In this example we will use the E.A.A. model to build a low-connectivity DCA model. \\\n",
    "The information that we have about the training RNA family consists in: the sequence alignment and the consenus secondary structure (both trough the Covariance Model) and the 3D contacts trough the PDB file. </font> \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "20924cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "include(\"../src/FCSeqTools.jl\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"3\">Here is an example of  RF00379 molecule and its associated consensus secondary structure. \\\n",
    "To make the execution faster we will not generate full lenght molecules but just a portion from nucleotide 55 to 102. </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "natural_sequences = do_number_matrix_prot(do_letter_matrix(\"../data/CM_130530_MC.fasta\"), 0.2);\n",
    "\n",
    "#Betalactamases_VIM-NDM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"3\">Here is a segment example with its associated secondary structure. \\\n",
    "The database has a different size because the data-cleaning procedure depends on the region selected. \\\n",
    "Now we will run the E.A.A. building up our ineraction netwotk edge by edge till we reach a good performance generative model. \\\n",
    "At each iteration the algorithm reports: the added edge, the iteration number, the number of total added edges and the connectivity percentace of the fully connected case.\\\n",
    "Each 15 iterations the algorithm reports: the model score (Pearson between natural and artificial two-point correlations), the model mean energy, the model partition function and the model entropy. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving data in:../training/CM_alignment/largest_component/three_different_pseudocounts/largest_component_reg=0.8_nb1\n",
      "Fully connected model has 4560 edges, 2010960 elements and a score around ~ 0.95\n",
      "\n",
      " iteration = 10    edges: 10,   elements: 10,   edge complexity: 0.22 %,  elements complexity: 0.0 %\n",
      "\n",
      " iteration = 200   edges: 125,   elements: 198,   edge complexity: 2.74 %,  elements complexity: 0.01 %\n",
      "\n",
      " iteration = 400   edges: 215,   elements: 392,   edge complexity: 4.71 %,  elements complexity: 0.02 %\n",
      "\n",
      " iteration = 600   edges: 294,   elements: 585,   edge complexity: 6.45 %,  elements complexity: 0.03 %\n",
      "\n",
      " iteration = 800   edges: 381,   elements: 770,   edge complexity: 8.36 %,  elements complexity: 0.04 %\n",
      "\n",
      " iteration = 1000   edges: 446,   elements: 958,   edge complexity: 9.78 %,  elements complexity: 0.05 %\n",
      "\n",
      " iteration = 1200   edges: 515,   elements: 1149,   edge complexity: 11.29 %,  elements complexity: 0.06 %\n",
      "\n",
      " iteration = 1400   edges: 576,   elements: 1336,   edge complexity: 12.63 %,  elements complexity: 0.07 %\n",
      "\n",
      " iteration = 1600   edges: 651,   elements: 1525,   edge complexity: 14.28 %,  elements complexity: 0.08 %\n",
      "\n",
      " iteration = 1800   edges: 710,   elements: 1716,   edge complexity: 15.57 %,  elements complexity: 0.09 %\n",
      "\n",
      " iteration = 2000   edges: 785,   elements: 1900,   edge complexity: 17.21 %,  elements complexity: 0.09 %\n",
      "\n",
      " iteration = 2200   edges: 824,   elements: 2087,   edge complexity: 18.07 %,  elements complexity: 0.1 %\n",
      "\n",
      " iteration = 2400   edges: 874,   elements: 2277,   edge complexity: 19.17 %,  elements complexity: 0.11 %\n",
      "\n",
      " iteration = 2600   edges: 922,   elements: 2459,   edge complexity: 20.22 %,  elements complexity: 0.12 %\n",
      "\n",
      " iteration = 2800   edges: 967,   elements: 2642,   edge complexity: 21.21 %,  elements complexity: 0.13 %\n",
      "\n",
      " iteration = 3000   edges: 1008,   elements: 2823,   edge complexity: 22.11 %,  elements complexity: 0.14 %\n",
      "\n",
      " iteration = 3200   edges: 1070,   elements: 3014,   edge complexity: 23.46 %,  elements complexity: 0.15 %\n",
      "\n",
      " iteration = 3400   edges: 1116,   elements: 3193,   edge complexity: 24.47 %,  elements complexity: 0.16 %\n",
      "\n",
      " iteration = 3600   edges: 1165,   elements: 3383,   edge complexity: 25.55 %,  elements complexity: 0.17 %\n",
      "\n",
      " iteration = 3800   edges: 1198,   elements: 3568,   edge complexity: 26.27 %,  elements complexity: 0.18 %\n",
      "\n",
      " iteration = 4000   edges: 1239,   elements: 3753,   edge complexity: 27.17 %,  elements complexity: 0.19 %\n",
      "\n",
      " iteration = 4200   edges: 1273,   elements: 3936,   edge complexity: 27.92 %,  elements complexity: 0.2 %\n",
      "\n",
      " iteration = 4400   edges: 1310,   elements: 4116,   edge complexity: 28.73 %,  elements complexity: 0.2 %\n",
      "\n",
      " iteration = 4600   edges: 1339,   elements: 4300,   edge complexity: 29.36 %,  elements complexity: 0.21 %\n",
      "\n",
      " iteration = 4800   edges: 1370,   elements: 4481,   edge complexity: 30.04 %,  elements complexity: 0.22 %\n",
      "\n",
      " iteration = 5000   edges: 1408,   elements: 4661,   edge complexity: 30.88 %,  elements complexity: 0.23 %\n",
      "\n",
      " iteration = 5200   edges: 1440,   elements: 4838,   edge complexity: 31.58 %,  elements complexity: 0.24 %\n",
      "\n",
      " iteration = 5400   edges: 1466,   elements: 5025,   edge complexity: 32.15 %,  elements complexity: 0.25 %\n",
      "\n",
      " iteration = 5600   edges: 1496,   elements: 5208,   edge complexity: 32.81 %,  elements complexity: 0.26 %\n",
      "\n",
      " iteration = 5800   edges: 1525,   elements: 5379,   edge complexity: 33.44 %,  elements complexity: 0.27 %\n",
      "\n",
      " iteration = 6000   edges: 1549,   elements: 5559,   edge complexity: 33.97 %,  elements complexity: 0.28 %\n",
      "\n",
      " iteration = 6200   edges: 1583,   elements: 5746,   edge complexity: 34.71 %,  elements complexity: 0.29 %\n",
      "\n",
      " iteration = 6400   edges: 1612,   elements: 5926,   edge complexity: 35.35 %,  elements complexity: 0.29 %\n",
      "\n",
      " iteration = 6600   edges: 1640,   elements: 6097,   edge complexity: 35.96 %,  elements complexity: 0.3 %\n",
      "\n",
      " iteration = 6800   edges: 1668,   elements: 6269,   edge complexity: 36.58 %,  elements complexity: 0.31 %\n",
      "\n",
      " iteration = 7000   edges: 1683,   elements: 6445,   edge complexity: 36.91 %,  elements complexity: 0.32 %\n",
      "\n",
      " iteration = 7200   edges: 1713,   elements: 6616,   edge complexity: 37.57 %,  elements complexity: 0.33 %\n",
      "\n",
      " iteration = 7400   edges: 1735,   elements: 6787,   edge complexity: 38.05 %,  elements complexity: 0.34 %\n",
      "\n",
      " iteration = 7600   edges: 1761,   elements: 6964,   edge complexity: 38.62 %,  elements complexity: 0.35 %\n",
      "\n",
      " iteration = 7800   edges: 1788,   elements: 7140,   edge complexity: 39.21 %,  elements complexity: 0.36 %\n",
      "\n",
      " iteration = 8000   edges: 1816,   elements: 7318,   edge complexity: 39.82 %,  elements complexity: 0.36 %\n",
      "\n",
      " iteration = 8200   edges: 1838,   elements: 7491,   edge complexity: 40.31 %,  elements complexity: 0.37 %\n",
      "\n",
      " iteration = 8400   edges: 1857,   elements: 7669,   edge complexity: 40.72 %,  elements complexity: 0.38 %\n",
      "\n",
      " iteration = 8600   edges: 1878,   elements: 7841,   edge complexity: 41.18 %,  elements complexity: 0.39 %\n",
      "\n",
      " iteration = 8800   edges: 1897,   elements: 8019,   edge complexity: 41.6 %,  elements complexity: 0.4 %\n",
      "\n",
      " iteration = 9000   edges: 1913,   elements: 8196,   edge complexity: 41.95 %,  elements complexity: 0.41 %\n",
      "\n",
      " iteration = 9200   edges: 1937,   elements: 8374,   edge complexity: 42.48 %,  elements complexity: 0.42 %\n",
      "\n",
      " iteration = 9400   edges: 1955,   elements: 8557,   edge complexity: 42.87 %,  elements complexity: 0.43 %\n",
      "\n",
      " iteration = 9600   edges: 1971,   elements: 8726,   edge complexity: 43.22 %,  elements complexity: 0.43 %\n",
      "\n",
      " iteration = 9800   edges: 1988,   elements: 8897,   edge complexity: 43.6 %,  elements complexity: 0.44 %\n"
     ]
    },
    {
     "ename": "LoadError",
     "evalue": "InterruptException:",
     "output_type": "error",
     "traceback": [
      "InterruptException:",
      "",
      "Stacktrace:",
      "  [1] Array",
      "    @ ./boot.jl:459 [inlined]",
      "  [2] Array",
      "    @ ./boot.jl:468 [inlined]",
      "  [3] Array",
      "    @ ./boot.jl:476 [inlined]",
      "  [4] similar",
      "    @ ./abstractarray.jl:841 [inlined]",
      "  [5] similar",
      "    @ ./abstractarray.jl:840 [inlined]",
      "  [6] similar",
      "    @ ./broadcast.jl:212 [inlined]",
      "  [7] similar",
      "    @ ./broadcast.jl:211 [inlined]",
      "  [8] copy",
      "    @ ./broadcast.jl:885 [inlined]",
      "  [9] materialize",
      "    @ ./broadcast.jl:860 [inlined]",
      " [10] gibbs_sampling(q::Int64, h_local::Vector{Float64}, Jij::Array{Float32, 3}, sequences::Matrix{Int8}, site_degree::Vector{Int64}, contact_list::Matrix{Int64}, sweeps::Int64)",
      "    @ Main ~/Documents/GitHub/FCSeqTools.jl/Example_Proteins/src/FCSeqTools.jl:168",
      " [11] (::var\"#6#10\"{IOStream, IOStream, IOStream, Int64, Int64, Float64, Int64, String, Float64, Float64, Int64, Vector{Any}, Float32, Vector{Float32}, Array{Float64, 3}, Vector{Float64}, Int64})(f3::IOStream)",
      "    @ Main ~/Documents/GitHub/FCSeqTools.jl/Example_Proteins/src/FCSeqTools.jl:530",
      " [12] open(::var\"#6#10\"{IOStream, IOStream, IOStream, Int64, Int64, Float64, Int64, String, Float64, Float64, Int64, Vector{Any}, Float32, Vector{Float32}, Array{Float64, 3}, Vector{Float64}, Int64}, ::String, ::Vararg{String}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})",
      "    @ Base ./io.jl:384",
      " [13] open",
      "    @ ./io.jl:381 [inlined]",
      " [14] (::var\"#5#9\"{IOStream, IOStream, Int64, Int64, Float64, Int64, String, Float64, Float64, Int64, Vector{Any}, Float32, Vector{Float32}, Array{Float64, 3}, Vector{Float64}, Int64})(f2::IOStream)",
      "    @ Main ~/Documents/GitHub/FCSeqTools.jl/Example_Proteins/src/FCSeqTools.jl:521",
      " [15] open(::var\"#5#9\"{IOStream, IOStream, Int64, Int64, Float64, Int64, String, Float64, Float64, Int64, Vector{Any}, Float32, Vector{Float32}, Array{Float64, 3}, Vector{Float64}, Int64}, ::String, ::Vararg{String}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})",
      "    @ Base ./io.jl:384",
      " [16] open",
      "    @ ./io.jl:381 [inlined]",
      " [17] (::var\"#4#8\"{IOStream, Int64, Int64, Float64, Int64, String, Float64, Float64, Int64, Vector{Any}, Float32, Vector{Float32}, Array{Float64, 3}, Vector{Float64}, Int64})(f1::IOStream)",
      "    @ Main ~/Documents/GitHub/FCSeqTools.jl/Example_Proteins/src/FCSeqTools.jl:520",
      " [18] open(::var\"#4#8\"{IOStream, Int64, Int64, Float64, Int64, String, Float64, Float64, Int64, Vector{Any}, Float32, Vector{Float32}, Array{Float64, 3}, Vector{Float64}, Int64}, ::String, ::Vararg{String}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})",
      "    @ Base ./io.jl:384",
      " [19] open",
      "    @ ./io.jl:381 [inlined]",
      " [20] (::var\"#3#7\"{Int64, Int64, Float64, Int64, String, Float64, Float64, Int64, Vector{Any}, Float32, Vector{Float32}, Array{Float64, 3}, Vector{Float64}, Int64})(f::IOStream)",
      "    @ Main ~/Documents/GitHub/FCSeqTools.jl/Example_Proteins/src/FCSeqTools.jl:519",
      " [21] open(::var\"#3#7\"{Int64, Int64, Float64, Int64, String, Float64, Float64, Int64, Vector{Any}, Float32, Vector{Float32}, Array{Float64, 3}, Vector{Float64}, Int64}, ::String, ::Vararg{String}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})",
      "    @ Base ./io.jl:384",
      " [22] open",
      "    @ ./io.jl:381 [inlined]",
      " [23] E_A_A(q::Int64, n_step::Int64, pseudo_count::Float64, number::Int64, number_matrix::Matrix{Int8}, filename::String, family_name::String, test_folder::String, method::String, notebook::String, fraction::Float64, stop::Float64)",
      "    @ Main ~/Documents/GitHub/FCSeqTools.jl/Example_Proteins/src/FCSeqTools.jl:518"
     ]
    }
   ],
   "source": [
    "using Random\n",
    "\n",
    "n_step = 100_000\n",
    "method = \"largest_component\"\n",
    "fraction = 0.1\n",
    "stop = 0.95\n",
    "pseudo_count = 0.8\n",
    "s = time()\n",
    "Random.seed!(2) \n",
    "notebook = \"nb1\"\n",
    "\n",
    "family_name = \"CM_alignment\" #\"Betalactamases\"\n",
    "test_folder = \"three_different_pseudocounts\"\n",
    "score, likelihood_gain, generated_sequences, Jij, h, contact_list, site_degree, edge_list, single_likelihood_gain_vector, path = E_A_A(21, n_step, pseudo_count, 12000, natural_sequences,\"example_output.txt\", family_name, test_folder, method, notebook, fraction, stop); \n",
    "\n",
    "s = time() - s\n",
    "\n",
    "# SAVE DATA ##################################################################################\n",
    "using JLD\n",
    "init_pseudo_count = 0.01\n",
    "#folder_name = \"../training/\" * method * string(fraction) * \"_stop=\" * string(stop) * \"_reg=\"*string(pseudo_count)*\"_h_ps-count=\" * string(init_pseudo_count) * \"nbook=\"*notebook\n",
    "JLD.save(path*\"/\"*\"score.jld\",\"data\", score)\n",
    "JLD.save(path*\"/\"*\"likelihood_gain.jld\",\"data\", likelihood_gain)\n",
    "JLD.save(path*\"/\"*\"generated_sequences.jld\",\"data\", generated_sequences)\n",
    "JLD.save(path*\"/\"*\"Jij.jld\",\"data\", Jij)\n",
    "JLD.save(path*\"/\"*\"h.jld\",\"data\", h)\n",
    "JLD.save(path*\"/\"*\"contact_list.jld\",\"data\", contact_list)\n",
    "JLD.save(path*\"/\"*\"site_degree.jld\",\"data\", site_degree)\n",
    "JLD.save(path*\"/\"*\"edge_list.jld\",\"data\", edge_list)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "13ebb2d5",
   "metadata": {},
   "source": [
    "## saving data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "909ba0e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "using JLD\n",
    "\n",
    "init_pseudo_count = 0.01\n",
    "folder_name = \"../training/\" * method * string(fraction) * \"_stop=\" * string(stop) * \"_reg=\"*string(pseudo_count)*\"_h_ps-count=\" * string(init_pseudo_count) * \"nbook=\"*notebook\n",
    "\n",
    "JLD.save(folder_name*\"/\"*\"score.jld\",\"data\", score)\n",
    "JLD.save(folder_name*\"/\"*\"likelihood_gain.jld\",\"data\", likelihood_gain)\n",
    "JLD.save(folder_name*\"/\"*\"generated_sequences.jld\",\"data\", generated_sequences)\n",
    "JLD.save(folder_name*\"/\"*\"Jij.jld\",\"data\", Jij)\n",
    "JLD.save(folder_name*\"/\"*\"h.jld\",\"data\", h)\n",
    "JLD.save(folder_name*\"/\"*\"contact_list.jld\",\"data\", contact_list)\n",
    "JLD.save(folder_name*\"/\"*\"site_degree.jld\",\"data\", site_degree)\n",
    "JLD.save(folder_name*\"/\"*\"edge_list.jld\",\"data\", edge_list)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0c4e1927",
   "metadata": {},
   "source": [
    "## Saving the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3919994c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "print_model_to_file_prot (generic function with 1 method)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function print_model_to_file_prot(natural_sequences,Jij,h,filename)\n",
    "    open(filename, \"w\") do f\n",
    "        for i in 1:length(natural_sequences[1,:])\n",
    "            for j in i+1:length(natural_sequences[1,:])\n",
    "                for k in 1:21    \n",
    "                    if k==1\n",
    "                        k2=1        \n",
    "                    elseif k==2\n",
    "                        k2=2\n",
    "                    elseif k==3\n",
    "                        k2=3\n",
    "                    elseif k==4\n",
    "                        k2=4\n",
    "                    elseif k==5\n",
    "                        k2=5\n",
    "                    elseif k==6\n",
    "                        k2=6\n",
    "                    elseif k==7\n",
    "                        k2=7\n",
    "                    elseif k==8\n",
    "                        k2=8\n",
    "                    elseif k==9\n",
    "                        k2=9\n",
    "                    elseif k==10\n",
    "                        k2=10\n",
    "                    elseif k==11\n",
    "                        k2=11\n",
    "                    elseif k==12\n",
    "                        k2=12\n",
    "                    elseif k==13\n",
    "                        k2=13\n",
    "                    elseif k==14\n",
    "                        k2=14\n",
    "                    elseif k==15\n",
    "                        k2=15\n",
    "                    elseif k==16\n",
    "                        k2=16\n",
    "                    elseif k==17\n",
    "                        k2=17\n",
    "                    elseif k==18\n",
    "                        k2=18\n",
    "                    elseif k==19\n",
    "                        k2=19\n",
    "                    elseif k==20\n",
    "                        k2=20\n",
    "                    elseif k==21\n",
    "                        k2=0\n",
    "                    end\n",
    "                        \n",
    "                    for l in 1:21\n",
    "                        if l==1\n",
    "                            l2=1\n",
    "                        elseif l==2\n",
    "                            l2=2\n",
    "                        elseif l==3\n",
    "                            l2=3\n",
    "                        elseif l==4\n",
    "                            l2=4\n",
    "                        elseif l==5\n",
    "                            l2=5\n",
    "                        elseif l==6\n",
    "                            l2=6\n",
    "                        elseif l==7\n",
    "                            l2=7\n",
    "                        elseif l==8\n",
    "                            l2=8\n",
    "                        elseif l==9\n",
    "                            l2=9\n",
    "                        elseif l==10\n",
    "                            l2=10\n",
    "                        elseif l==11\n",
    "                            l2=11\n",
    "                        elseif l==12\n",
    "                            l2=12\n",
    "                        elseif l==13\n",
    "                            l2=13\n",
    "                        elseif l==14\n",
    "                            l2=14\n",
    "                        elseif l==15\n",
    "                            l2=15\n",
    "                        elseif l==16\n",
    "                            l2=16\n",
    "                        elseif l==17\n",
    "                            l2=17\n",
    "                        elseif l==18\n",
    "                            l2=18\n",
    "                        elseif l==19\n",
    "                            l2=19\n",
    "                        elseif l==20\n",
    "                            l2=20\n",
    "                        elseif l==21\n",
    "                            l2=0\n",
    "                        end\n",
    "                        opo=Jij[i,j,21*(k-1)+l]\n",
    "                        write(f,\"\\nJ $(i-1) $(j-1) $(k2) $(l2) $opo\" )\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        for i in 1:length(natural_sequences[1,:])\n",
    "            for j in 1:21          \n",
    "                if j==1\n",
    "                    j2=1\n",
    "                elseif j==2\n",
    "                    j2=2\n",
    "                elseif j==3\n",
    "                    j2=3\n",
    "                elseif j==4\n",
    "                    j2=4\n",
    "                elseif j==5\n",
    "                    j2=5\n",
    "                elseif j==6\n",
    "                    j2=6\n",
    "                elseif j==7\n",
    "                    j2=7\n",
    "                elseif j==8\n",
    "                    j2=8\n",
    "                elseif j==9\n",
    "                    j2=9\n",
    "                elseif j==10\n",
    "                    j2=10\n",
    "                elseif j==11\n",
    "                    j2=11\n",
    "                elseif j==12\n",
    "                    j2=12\n",
    "                elseif j==13\n",
    "                    j2=13\n",
    "                elseif j==14\n",
    "                    j2=14\n",
    "                elseif j==15\n",
    "                    j2=15\n",
    "                elseif j==16\n",
    "                    j2=16\n",
    "                elseif j==17\n",
    "                    j2=17\n",
    "                elseif j==18\n",
    "                    j2=18\n",
    "                elseif j==19\n",
    "                    j2=19\n",
    "                elseif j==20\n",
    "                    j2=20\n",
    "                elseif j==21\n",
    "                    j2=0\n",
    "                end\n",
    "                opo=h[21*(i-1)+j]\n",
    "                write(f,\"\\nh $(i-1) $(j2) $opo\" )\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1c9c8dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"../saved_models/model_cumulative0.3_pseudocount0.5_init_pscount0.01_nbook_test1.txt\"\n",
    "print_model_to_file_prot(natural_sequences, Jij, h, filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"3\">The model obtained has a performance comparable to the fully connected DCA while having just ~20% of its connectivity. The entropy of the model is 35.08. This means that it is able to generate e³⁵ (3.5x10¹⁵) different 55-102 segments for the RF00379 family. \\\n",
    "Now we can test our artificial sequences. We do the classical statistical check of the PCA projection and the two-point correlation representation. \\\n",
    "We test the performance of our model against the one of the Covariance Model. The CM model only contains trivial one-point and secondary information so our model must do better than it. </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cm_sequences = rna_cm_model_generation(0.8,0.05,7000,natural_sequences,ss_contact_matrix);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_stat_check(natural_sequences, generated_sequences, cm_sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"3\">The E.A.A. artificial molecules are practically statistically indistinguishable from the natural ones. We see that they have a very similar PCA projection (artificial one seems richer just because we have more artificial sequences than natural ones) while Covariance Model fails to capture the details of the distribution. \n",
    "    The selected model has almost a perfect two-point statistics for all site pairs while the CM model only captures it for the ones involved in secondary structure contacts. \\\n",
    "     </font>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"3\">The interpretability is one of the main reasons in our quest to find parsimonious generative models. Now that we are sure we obtained a good generative model with relatively few parameters we can try to interprete them. \\\n",
    "Dividing the added edges in secondary structure contacts, 3D contacts we have:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_interpretation_plot(len,ss_contact_matrix,tertiary_contact_matrix,edge_list[1:50,:])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"3\">We see that the secondary structure contacts are taken in the first few iteration. We have lot of neighbouring sites probably due to philogenic effects. It is striking that we see some 3D contacts (in particular around site 40) before the NONE edges. This\n",
    "suggests that our algorithm effectively captures some information about the tertiary structure. \\\n",
    "Those results, that are far more general than this simple example, suggest that the added edges have a co-evolutionary interpretation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"3\">\n",
    "This notebook serves as an example of the application of the techniques described in the main text.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.8.5",
   "language": "julia",
   "name": "julia-1.8"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.8.5"
  },
  "vscode": {
   "interpreter": {
    "hash": "6412a36ae9b8109c1adf793ae61659867359f52a3cfb701a53c331fa90dc1d37"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
